src/tasks/run_closed_loop.py

```diff
@@ -16,15 +56,33 @@ def run():
         print('pip install "neural-data-simulator[extras]"')
         return
 
+    initialize_logger(SCRIPT_NAME)
+    configure_logger(SCRIPT_NAME, LogLevel._INFO)
+
+    logger.info("Starting modules")
     encoder = subprocess.Popen(["encoder"])
     ephys = subprocess.Popen(["ephys_generator"])
     decoder = subprocess.Popen(["decoder"])
-    center_out_reach = subprocess.Popen(["center_out_reach"])
 
-    center_out_reach.wait()
-    encoder.kill()
-    ephys.kill()
-    decoder.kill()
+    temp_dir, center_out_reach_pipe_path = _create_pipe_file("center_out_reach")
+    center_out_reach = subprocess.Popen(
+        ["center_out_reach", "--pipe", center_out_reach_pipe_path]
+    )
+    logger.info("Modules started")
+
+    main_task_finished = _wait_for_main_task_finish(center_out_reach_pipe_path)
+    _terminate_process("encoder", encoder)
+    _terminate_process("ephys_generator", ephys)
+    _terminate_process("decoder", decoder)
+    if not main_task_finished:
+        _terminate_process("center_out_reach", center_out_reach)
+    else:
+        logger.info("Waiting for center_out_reach")
+        center_out_reach.wait()
+
+    os.remove(center_out_reach_pipe_path)
+    os.rmdir(temp_dir)
```

Comment:
you can remove this if you use tempfile with context managers:
https://docs.python.org/3/library/tempfile.html#tempfile-examples

Reply:
I saw this function before, but the TemporaryFile() already creates a temporary file, but that file is _not_ a pipe file. We need to call the [mkfifo](https://docs.python.org/3/library/os.html#os.mkfifo) function to create one.

Reply:
so maybe the best solution here would be for you to create the context, in this way, you can be sure that os.remove and os.rmdir will be called even if there's an exception in the code

Reply:
Using context manager now! 🎉